[Heng Yi Qun] iP

hengyiqun opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

hengyiqun committed:
Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

hengyiqun committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

hengyiqun committed:
Added a new class called Task.
Added the ability to mark tasks as done.

hengyiqun committed:
Add support for tracking three types of tasks:

1. ToDos: tasks without any date/time attached to it e.g., visit new theme park
2. Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
3. Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

hengyiqun committed:
Updated Event, Todo and Deadline so that the type of "done" task would show up when "list" is called (i.e. [E], [D] or [T])

hengyiqun committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

hengyiqun committed:
Teach Duke to deal with errors such as incorrect inputs entered by the user.
Errors included are Insufficient Argument for Todo, Deadline, Event and Done, as well as Wrong Argument (when an invalid keyword is used).

hengyiqun committed:
Add support for deleting tasks from the list.
Add Insufficient Argument Exception for "delete"

hengyiqun committed:
Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up.

hengyiqun committed:
Store deadline dates as a java.time.LocalDate in my task objects. Accept dates in a format of yyyy-mm-dd format (e.g., 2019-10-15) and print in a different format of MMM dd yyyy e.g., (Oct 15 2019).

hengyiqun committed:
Merge branch 'branch-level-7'

hengyiqun committed:
Merge branch 'branch-Level-8'

Resolve Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

hengyiqun committed:
Addthe following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list

hengyiqun committed:
Add JavaDoc header comments to the code.

hengyiqun committed:
Add JUnit tests to test the behavior of the code.

hengyiqun committed:
Add more javadocs comments

hengyiqun committed:
Tweak the code slightly to comply with the given coding standard in CS2103

hengyiqun committed:
Give users a way to find a task by searching for a keyword.

hengyiqun committed:
Merge branch 'branch-A-JavaDoc'

hengyiqun committed:
Merge branch 'branch-A-CodingStandard'

hengyiqun committed:
Merge branch 'branch-Level-9'

wangtao0717 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Constructs a deadline that has yet to be completed
+     * @param description description of the deadline
+     * @param time due date in yyyy-MM-dd format (e.g. 2021-03-12)
```
According to code standard, I think leaving a line break between the description and parameter section will be better.

wangtao0717 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+            try {
+                FileWriter fw = new FileWriter(this.filePath);
+                for (Task t : this.tasks.getList()) {
+                    String rmb = "";
```
Perhaps a more intuitive variable name here?

jasaaanlim replied:
Even though 'rmb' is just a temporary short-lived variable, would a more readable variable name make things easier for other collaborators and when looking back in future?

wangtao0717 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+                    store.add(todo);
+                    System.out.println("     " + todo);
+                    System.out.println("     Now you have " + store.size() + " tasks in the list.");
+                } else if (parts[0].equals("deadline")) {
```
Perhaps it may be more concise to replace if-else if method with switch method here.

wangtao0717 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+            }
+        }
+    }*/
+}
```
The Duke.java seems too long, will it be better to extracte some code out as some classes or functions?

jasaaanlim replied:
I agree with @wangtao0717 on the Duke.java being a little too long, maybe you could look into that.

wangtao0717 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,190 @@
+                return new Parser("list", null, null, -1);
+            } else if (parts[0].equals("done")) {
+                if (parts.length == 1) {
+                    throw new InsufficientArgumentsException("     ☹ OOPS!!! The " +
```
Perhaps it is better to break before an operator "+"

wangtao0717 commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,147 @@
+            int count = 1;
+            for (Task t : taskList) {
+                if (count == taskList.size()) {
+                    msg += "     " + count + "." + t.toString();
```
There is a lot of the same indentation, maybe it can be extracted into a variable, which can be changed easily later？

wangtao0717 reviewed (commented):
In general, there is no major problem in code quality. The code is clear and easy to understand. If it can replace some long if-elseif methods with the switch method, the code may be more concise. And I think it would be better if some parts are not repeated as often, which can be extracted out to some classes or methods.

jasaaanlim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+                        rmb += "1";
+                    } else {
+                        rmb += "0";
+                    }
```
Could a switch case be a neater alternative as compared to a long list of if-else?

jasaaanlim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+                        rmb += "D";
+                    } else if (t instanceof Todo) {
+                        rmb += "T";
+                    }
```
I noticed a few repetitions throughout the various classes where you had to use multiple if else statements to display the event type. Have you thought about handling the display of an event type in the respective toString() methods of the Task? Would that be a more efficient way?

jasaaanlim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+            while (fileScanner.hasNext()) {
+                String curRecTask = fileScanner.nextLine();
+                String[] items = curRecTask.split(" \\| ");
+                boolean done = (items[1].equals("1") ? true : false);
```
Noticed that a few of your boolean variables were not named to sound like boolean variables. E.g. the 'first' and 'done' booleans. In this case, would it be more intuitive to name this boolean variable as isDone instead?

jasaaanlim commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,190 @@
+                return new Parser("find", toFind, null, -1);
+            } else if (parts[0].equals("todo")) {
+                if (parts.length == 1) {
+                    throw new InsufficientArgumentsException("     ☹ OOPS!!! The " +
```
Is there a reason why you are hardcoding the tabs? Would '\t' may be a neater way of displaying your tabs?

jasaaanlim commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,147 @@
+            String msg = "     Got it. I've added this task:\n";
+            LocalDate dueDate = parser.getDueDate();
+            String taskDescription = parser.getDescription();
+            Deadline deadline = new Deadline(taskDescription, dueDate);
```
Since you do not plan to use dueDate or taskDescription, could you directly place parser.getDueDate() and parser.getDescription() in the constructor instead?

jasaaanlim reviewed (commented):
The program is generally well written with some minor issues with the naming conventions and there can be some improvements on reusing to code and variable reuse.

tlylt commented on src/main/java/DueDate.java:
```diff
@@ -0,0 +1,6 @@
+/**
+ * DueDate is an interface for inheritance by tasks with a due date
+ */
+public interface DueDate {
```
Good job on naming the interface correctly in PascalCase. I like your use of an interface here.

tlylt commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+ * Duke allows the user to maintain a list of tasks
+ */
 public class Duke {
+
```
Hi, just a small inconsistency that I spotted. For some of your classes, there is an empty line between the class declaration and the class properties. However, some of the classes don't. 

Perhaps it will be more consistent if all classes have an empty line in between or none of them have an empty line in between?

tlylt commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,291 @@
+                                "description of an event cannot be empty.");
+                    }
+                    System.out.println("     Got it. I've added this task:");
+                    boolean flag = false, first = true;
```
Perhaps a more descriptive name for the boolean variable (instead of "flag") will be better?

tlylt reviewed (commented):
Overall good code! Just a few nits to fix.

hengyiqun committed:
Merge remote-tracking branch 'origin/add-gradle-support'

hengyiqun committed:
Use checkStyle to detect coding style violations.

hengyiqun committed:
Update file name
Update file path in build.gradle

hengyiqun committed:
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

hengyiqun committed:
Merge branch 'branch-Level-10'

hengyiqun committed:
Divide classes into packages

hengyiqun committed:
Add various command files to incorporate SLAP in Ui

hengyiqun committed:
Refractor to simplify Parser by including short methods in various commands

hengyiqun committed:
Use assert feature to ensure that the TaskList is properly updated whenever a relevant command is called by the user
Fix bugs associated with counting the number of tasks for EventCommand, DeadlineCommad and TodoCommand

hengyiqun committed:
Rename variables in Storage to fit coding standards

hengyiqun committed:
Merge pull request #2 from hengyiqun/branch-A-Assertions

Use assert feature to ensure that the TaskList is properly updated wh…

hengyiqun committed:
Merge pull request #3 from hengyiqun/branch-A-CodeQuality

Rename variables in Storage to fit coding standards

hengyiqun committed:
Merge branch 'master' of https://github.com/hengyiqun/ip

hengyiqun committed:
Add the ability to recognise and deal with duplicate items

hengyiqun committed:
Add more javadocs comments, mostly to the classes in the tasks package.

hengyiqun committed:
Add a User Guide to the project

hengyiqun committed:
Set theme jekyll-theme-cayman

hengyiqun committed:
Set theme jekyll-theme-cayman

hengyiqun committed:
Resize DialogBox according to the size of text input and the size of duke response

hengyiqun committed:
Merge branch 'master' of https://github.com/hengyiqun/ip

hengyiqun committed:
Add @Override tags to relevant methods in the commands package, and add header comments to the files in the command pacakage.

hengyiqun committed:
Set up packaging for unit tests.

hengyiqun committed:
Format EventCommad.

hengyiqun committed:
Update mainClassName in build.gradle for the jar file to launch successfully.

hengyiqun committed:
Change file name Ui.PNG to Ui.png to match grading script requirements.

hengyiqun committed:
Add DateCommand so as to prevent non date related commands from inheriting the method extractDate().
Add exception handling when the user keys in invalid date related commands.

hengyiqun committed:
Add error handling for cases when the user indicates invalid dates.

hengyiqun committed:
Add exception handling when a user inputs an invalid number or a string as the task that is completed or that is to be deleted.

hengyiqun committed:
Add ability to save tasks on local PC.

hengyiqun committed:
Add ability to save tasks on local PC.

hengyiqun committed:
Add exception handling when the user enters the command to delete/done a non-positive task.
